clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -2055,83 +2030,110 @@ Expected<SmallVector<StringRef>> linkAndWrapDeviceFiles(
     auto LinkerArgs = getLinkerArgs(Input, BaseArgs);
     DenseSet<OffloadKind> ActiveOffloadKinds;
     bool HasSYCLOffloadKind = false;
+    bool HasNonSYCLOffloadKinds = false;
     for (const auto &File : Input) {
       if (File.getBinary()->getOffloadKind() != OFK_None)
         ActiveOffloadKinds.insert(File.getBinary()->getOffloadKind());
       if (File.getBinary()->getOffloadKind() == OFK_SYCL)
         HasSYCLOffloadKind = true;
+      else
+        HasNonSYCLOffloadKinds = true;
     }
-
-    // First link and remove all the input files containing bitcode.
-    SmallVector<StringRef> InputFiles;
-    if (Error Err = linkBitcodeFiles(Input, InputFiles, LinkerArgs))
-      return Err;
-
-    // Write any remaining device inputs to an output file for the linker.
-    for (const OffloadFile &File : Input) {
-      auto FileNameOrErr = writeOffloadFile(File);
-      if (!FileNameOrErr)
-        return FileNameOrErr.takeError();
-      InputFiles.emplace_back(*FileNameOrErr);
-    }
-
     if (HasSYCLOffloadKind) {
-      // Link the remaining device files using the device linker for SYCL
+      SmallVector<StringRef> InputFiles;
```

Comment:
Design so far worked well for SYCL + SPIRV target. For non-spirv targets, we need to call clang instead of spirv-translator+AOTCompile on each of the split file output of sycl-post-link. Hence sycl-post-link and parsing of it table output has been moved here.
Initial validation has been performed with E2E tests on local third party hardware.